[Permission] permission presentation 계층 구현

vwx12345 opened the pull request:
### 💻 작업 내용
- permission 도메인의 presentation 계층을 구현했습니다.
- PermissionFakeController 에는 Profile 어노테이션을 통해 dev 환경일 때만 Bean 등록되게 설정했습니다.

### ✨ 리뷰 포인트
- 

### 📝 메모
- 이벤트로 처리될 Service들을 호출해보기 위해 FakeController를 생성했습니다.
- event 처리를 위해 tmp 디렉토리 아래에 Event 객체들을 생성했습니다.

### 🎯 관련 이슈
closed #49

vwx12345 committed:
feat(permission): permission presentation 계층 구현

vwx12345 committed:
feat(permission): 테스트용 PermissionFakeController 추가

- 테스트 환경에서 권한 생성 및 삭제를 위한 컨트롤러 추가
- /test/permissions/boards/{boardId}/create (userId 파라미터 필요)
- /test/permissions/boards/{boardId}/delete (userId 파라미터 필요)

vwx12345 committed:
feat(permission): DelegateRoleConverter 추가 및 적용

vwx12345 committed:
docs(permission): 주석 작성

vwx12345 committed:
feat(permission): eventHandler 구현

vwx12345 committed:
test(permission): test 구현

codingbaraGo commented on src/main/java/com/simpleboard/board/permission/presentation/PermissionController.java:
```diff
@@ -0,0 +1,28 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/permissions")
+public class PermissionController {
+  private final PermissionCommandService permissionCommandService;
+
+  @PostMapping("/boards/{boardId}/delegate")
+  public ResponseEntity<Void> delegateRole(
+      @PathVariable Long boardId, @RequestBody DelegateRoleForm form) {
```
Swagger 관련 어노테이션은 사용하지 않으시나요?
나중에 추가 예정인가요?

vwx12345 replied:
> Swagger 관련 어노테이션은 사용하지 않으시나요? 나중에 추가 예정인가요?

미처 신경을 못썼습니다. 추가해서 다시 pr 올리겠습니다.

codingbaraGo commented on src/main/java/com/simpleboard/board/permission/presentation/PermissionFakeController.java:
```diff
@@ -0,0 +1,54 @@
+ *
+ * @domain adapter-in
+ */
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/test/permissions")
+public class PermissionFakeController {
+  private final PermissionCommandService permissionCommandService;
+
```
이 클래스는 나중에 삭제될 컨트롤러인가요?
테스트라는게 서버를 띄워서 api를 호출해보는 테스트를 말하는건가요?
어떤 관점에서 생성한 테스트 컨트롤러인지 궁금하네요.
서버를 띄워서 api 호출을 해보는 컨트롤러라면 그냥 local에서만 가지고 있는게 좋을것 같구요,
다른 테스트에서 사용되는 컨트롤러라면 폴더를 test/ 하위로 옮기는게 좋을것같아요.

vwx12345 replied:
> 이 클래스는 나중에 삭제될 컨트롤러인가요? 테스트라는게 서버를 띄워서 api를 호출해보는 테스트를 말하는건가요? 어떤 관점에서 생성한 테스트 컨트롤러인지 궁금하네요. 서버를 띄워서 api 호출을 해보는 컨트롤러라면 그냥 local에서만 가지고 있는게 좋을것 같구요, 다른 테스트에서 사용되는 컨트롤러라면 폴더를 test/ 하위로 옮기는게 좋을것같아요.

이 클래스는 이벤트 기반으로 동작하는 서비스 로직을 보다 간편하게 테스트하기 위해 만든 임시 컨트롤러입니다.
원래 흐름대로라면 특정 이벤트를 발행해야 서비스가 동작하는데, 이 과정을 재현하려면 이벤트 객체를 직접 만들어 발행해야 하고, 이는 반복적으로 작성하기 번거롭습니다.

그래서 로컬 환경에서 Swagger나 간단한 API 호출만으로 이벤트를 트리거할 수 있도록 PermissionFakeController를 만들었습니다.
말씀하신대로 운영 환경에는 노출할 필요가 없으므로 local 프로필에서만 활성화하는 방식으로 처리하는 건 어떻게 생각하시나요?

codingbaraGo replied:
> > 이 클래스는 나중에 삭제될 컨트롤러인가요? 테스트라는게 서버를 띄워서 api를 호출해보는 테스트를 말하는건가요? 어떤 관점에서 생성한 테스트 컨트롤러인지 궁금하네요. 서버를 띄워서 api 호출을 해보는 컨트롤러라면 그냥 local에서만 가지고 있는게 좋을것 같구요, 다른 테스트에서 사용되는 컨트롤러라면 폴더를 test/ 하위로 옮기는게 좋을것같아요.
> 
> 이 클래스는 이벤트 기반으로 동작하는 서비스 로직을 보다 간편하게 테스트하기 위해 만든 임시 컨트롤러입니다. 원래 흐름대로라면 특정 이벤트를 발행해야 서비스가 동작하는데, 이 과정을 재현하려면 이벤트 객체를 직접 만들어 발행해야 하고, 이는 반복적으로 작성하기 번거롭습니다.
> 
> 그래서 로컬 환경에서 Swagger나 간단한 API 호출만으로 이벤트를 트리거할 수 있도록 PermissionFakeController를 만들었습니다. 말씀하신대로 운영 환경에는 노출할 필요가 없으므로 local 프로필에서만 활성화하는 방식으로 처리하는 건 어떻게 생각하시나요?

local 프로필에서만 활성화 하는 방법 좋네요.

vwx12345 replied:
> > > 이 클래스는 나중에 삭제될 컨트롤러인가요? 테스트라는게 서버를 띄워서 api를 호출해보는 테스트를 말하는건가요? 어떤 관점에서 생성한 테스트 컨트롤러인지 궁금하네요. 서버를 띄워서 api 호출을 해보는 컨트롤러라면 그냥 local에서만 가지고 있는게 좋을것 같구요, 다른 테스트에서 사용되는 컨트롤러라면 폴더를 test/ 하위로 옮기는게 좋을것같아요.
> > 
> > 
> > 이 클래스는 이벤트 기반으로 동작하는 서비스 로직을 보다 간편하게 테스트하기 위해 만든 임시 컨트롤러입니다. 원래 흐름대로라면 특정 이벤트를 발행해야 서비스가 동작하는데, 이 과정을 재현하려면 이벤트 객체를 직접 만들어 발행해야 하고, 이는 반복적으로 작성하기 번거롭습니다.
> > 그래서 로컬 환경에서 Swagger나 간단한 API 호출만으로 이벤트를 트리거할 수 있도록 PermissionFakeController를 만들었습니다. 말씀하신대로 운영 환경에는 노출할 필요가 없으므로 local 프로필에서만 활성화하는 방식으로 처리하는 건 어떻게 생각하시나요?
> 
> local 프로필에서만 활성화 하는 방법 좋네요.

그럼 회의 때 한번 논의해보죠!

codingbaraGo commented on src/main/java/com/simpleboard/board/permission/presentation/converter/DelegateRoleConverter.java:
```diff
@@ -0,0 +1,19 @@
+import com.simpleboard.board.permission.application.command.dto.DelegateRoleCommand;
+import com.simpleboard.board.permission.presentation.dto.DelegateRoleForm;
+
+public class DelegateRoleConverter {
+
+  private DelegateRoleConverter() {
+    throw new UnsupportedOperationException("Utility class");
+  }
+
+  public static DelegateRoleCommand toCommand(DelegateRoleForm form) {
+    return new DelegateRoleCommand(
```
저도 항상 utility 클래스를 정의할때 고민하는 부분인데요,
메서드들을 전부 static으로 선언하는것과, @Component 어노테이션을 붙이고 싱글톤으로 관리하는 것 중 어떤게 더 좋을까요?
static으로 사용하는것을 선택하신 이유가 있나요?

vwx12345 replied:
> 저도 항상 utility 클래스를 정의할때 고민하는 부분인데요, 메서드들을 전부 static으로 선언하는것과, @component 어노테이션을 붙이고 싱글톤으로 관리하는 것 중 어떤게 더 좋을까요? static으로 사용하는것을 선택하신 이유가 있나요?

해당 클래스는 상태도 없고 외부 의존성도 없는 순수 유틸성 로직이어서 static 메서드로 구현했습니다.
단순 유틸을 @Component로 관리하면 스프링 컨텍스트 의존이 생기고 불필요한 빈 생성·주입 비용이 발생하는 단점이 있어, 이런 경우에는 static이 더 적합하다고 판단했습니다.
반대로 DI가 필요하거나 모킹 가능성이 보인다면, @Component로 등록해 싱글턴 빈으로 관리하는 편이 더 좋다고 생각합니다.

codingbaraGo replied:
> > 저도 항상 utility 클래스를 정의할때 고민하는 부분인데요, 메서드들을 전부 static으로 선언하는것과, @component 어노테이션을 붙이고 싱글톤으로 관리하는 것 중 어떤게 더 좋을까요? static으로 사용하는것을 선택하신 이유가 있나요?
> 
> 해당 클래스는 상태도 없고 외부 의존성도 없는 순수 유틸성 로직이어서 static 메서드로 구현했습니다. 단순 유틸을 @component로 관리하면 스프링 컨텍스트 의존이 생기고 불필요한 빈 생성·주입 비용이 발생하는 단점이 있어, 이런 경우에는 static이 더 적합하다고 판단했습니다. 반대로 DI가 필요하거나 모킹 가능성이 보인다면, @component로 등록해 싱글턴 빈으로 관리하는 편이 더 좋다고 생각합니다.

static이 효율성이 더 좋긴 하겠네요.
다만, 빈 생성과 DI에 발생하는 비용은 서버를 올릴때만 발생하는 비용이고, 비용이 크지 않으니 다른 부분도 고려해보죠.

일단 의존성의 경우 도메인 계층 converter라면 고민의 여지가 있겠지만, presentation 계층이니 spring에 대한 의존성은 있어도 괜찮다고 생각합니다.

또, 코드의 통일성도 중요한 요소 중 하나라고 생각합니다.
같은 converter인데 어떤 클래스는 static으로 사용해야하고, 다른 클래스는 DI를 써야하면 유지보수 하는 입장에서 쓸때마다 어떻게 써야 하는지 확인을 해야겠죠. (좋은 코드는 개발자가 바뀌어도 부드러운 유지보수가 가능한 코드니까요.)
그러면 모든 converter를 static으로 사용할 수 있냐가 관건이겠네요.
만약 스프링 빈으로 등록해야하는 converter가 있다면 나머지도 스프링 빈으로 등록하는게 유지보수 측면에서 좋지 않을까 생각해요.

그렇다면 저희는
1. 지금 시점에서 스프링 빈으로 등록해야하는 converter가 존재할까?
2. 향후 기능 추가를 진행하며 추가될 converter중 스프링 빈으로 등록해야 할 converter가 있을까?

두가지 요소를 고민해봐야겠네요.

codingbaraGo replied:
> 
> static이 효율성이 더 좋긴 하겠네요. 다만, 빈 생성과 DI에 발생하는 비용은 서버를 올릴때만 발생하는 비용이고, 비용이 크지 않으니 다른 부분도 고려해보죠.
> 
> 일단 의존성의 경우 도메인 계층 converter라면 고민의 여지가 있겠지만, presentation 계층이니 spring에 대한 의존성은 있어도 괜찮다고 생각합니다.
> 
> 또, 코드의 통일성도 중요한 요소 중 하나라고 생각합니다. 같은 converter인데 어떤 클래스는 static으로 사용해야하고, 다른 클래스는 DI를 써야하면 유지보수 하는 입장에서 쓸때마다 어떻게 써야 하는지 확인을 해야겠죠. (좋은 코드는 개발자가 바뀌어도 부드러운 유지보수가 가능한 코드니까요.) 그러면 모든 converter를 static으로 사용할 수 있냐가 관건이겠네요. 만약 스프링 빈으로 등록해야하는 converter가 있다면 나머지도 스프링 빈으로 등록하는게 유지보수 측면에서 좋지 않을까 생각해요.
> 
> 그렇다면 저희는
> 
> 1. 지금 시점에서 스프링 빈으로 등록해야하는 converter가 존재할까?
> 2. 향후 기능 추가를 진행하며 추가될 converter중 스프링 빈으로 등록해야 할 converter가 있을까?
> 
> 두가지 요소를 고민해봐야겠네요.

https://github.com/value 어노테이션을 사용하는 converter나 repository, service 클래스를 사용하는 converter가 존재한다면 스프링 빈으로 등록해야겠군요

codingbaraGo commented on src/main/java/com/simpleboard/board/permission/presentation/PermissionEventHandler.java:
```diff
@@ -0,0 +1,28 @@
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class PermissionEventHandler {
+
+  private final PermissionCommandService permissionCommandService;
```
이벤트 핸들러의 경우에는 향후 기술 교체가 이루어질 확률이 높으니 인터페이스로 정의해두는건 어떨까요?

vwx12345 replied:
> 이벤트 핸들러의 경우에는 향후 기술 교체가 이루어질 확률이 높으니 인터페이스로 정의해두는건 어떨까요?

네, 그렇게 하겠습니다. 이벤트 핸들러는 향후 기술 교체 가능성을 고려해 인터페이스로 정의하겠습니다.

codingbaraGo reviewed (commented):
드디어 presentation 계층까지 나왔군요.
고생하셨습니다.

vwx12345 committed:
feat(permission): PermissionEventHandler 인터페이스 추가 및 구현체 분리

- PermissionEventHandler 인터페이스 생성
- 기존 PermissionEventHandler를 PermissionEventHandlerImpl로 변경

vwx12345 committed:
feat(permission): swagger 설정 추가

codingbaraGo reviewed (approved):
고생하셨습니다.

vwx12345 committed:
refactor(permission): Profile 적용

- dev 환경에서만 bean에 등록될 수 있게 @Profile("dev") 추가

codingbaraGo commented on src/main/java/com/simpleboard/board/permission/presentation/PermissionFakeController.java:
```diff
@@ -35,6 +36,7 @@
 @RequiredArgsConstructor
 @RestController
 @RequestMapping("/test/permissions")
+@Profile("dev")
```
저는 fake controller를 로컬 환경에서만 돌린다고 생각했는데, local이 아니라 dev 맞나요?

vwx12345 replied:
> 저는 fake controller를 로컬 환경에서만 돌린다고 생각했는데, local이 아니라 dev 맞나요?

저는 prod만 아니면 될거라고 생각했는데 @Profile("!prod")로 변경하는 건 어떻게 생각하시나요?

codingbaraGo replied:
> > 저는 fake controller를 로컬 환경에서만 돌린다고 생각했는데, local이 아니라 dev 맞나요?
> 
> 저는 prod만 아니면 될거라고 생각했는데 @Profile("!prod")로 변경하는 건 어떻게 생각하시나요?

이 전 버전에서는 FakeController가 어떤 용도의 컨트롤러인지 모르겠었는데 지금 버전을 보니 테스트용 어드민 api같군요.
이런 어드민들이 앞으로 더 필요할까요?
아예 별도의 패키지로 모으는게 좋지 않을까 해서요.
이런 어드민 컨트롤러들은 스테이징과 릴리즈에서는 아예 빌드 대상에서 빠져야한다 생각합니다.

codingbaraGo replied:
그리고 프로파일을 dev로 한정해놓으면 로컬 프로파일 활성화시 사용하지 못할것같아요.
프로파일 설정은 !prod가 맞다고 생각합니다.

vwx12345 replied:
> > > 저는 fake controller를 로컬 환경에서만 돌린다고 생각했는데, local이 아니라 dev 맞나요?
> > 
> > 
> > 저는 prod만 아니면 될거라고 생각했는데 @Profile("!prod")로 변경하는 건 어떻게 생각하시나요?
> 
> 이 전 버전에서는 FakeController가 어떤 용도의 컨트롤러인지 모르겠었는데 지금 버전을 보니 테스트용 어드민 api같군요. 이런 어드민들이 앞으로 더 필요할까요? 아예 별도의 패키지로 모으는게 좋지 않을까 해서요. 이런 어드민 컨트롤러들은 스테이징과 릴리즈에서는 아예 빌드 대상에서 빠져야한다 생각합니다.

당장은 prod 제외 프로필(!prod) 에서만 동작하도록 두고, 추후에는 이런 어드민/테스트 컨트롤러가 늘어 난다면 별도 소스셋으로 모아 운영 JAR에서 완전히 배제하는 방식으로 전환하는게 좋을 것 같습니다.

codingbaraGo replied:
> > > > 저는 fake controller를 로컬 환경에서만 돌린다고 생각했는데, local이 아니라 dev 맞나요?
> > > 
> > > 
> > > 저는 prod만 아니면 될거라고 생각했는데 @Profile("!prod")로 변경하는 건 어떻게 생각하시나요?
> > 
> > 
> > 이 전 버전에서는 FakeController가 어떤 용도의 컨트롤러인지 모르겠었는데 지금 버전을 보니 테스트용 어드민 api같군요. 이런 어드민들이 앞으로 더 필요할까요? 아예 별도의 패키지로 모으는게 좋지 않을까 해서요. 이런 어드민 컨트롤러들은 스테이징과 릴리즈에서는 아예 빌드 대상에서 빠져야한다 생각합니다.
> 
> 당장은 prod 제외 프로필(!prod) 에서만 동작하도록 두고, 추후에는 이런 어드민/테스트 컨트롤러가 늘어 난다면 별도 소스셋으로 모아 운영 JAR에서 완전히 배제하는 방식으로 전환하는게 좋을 것 같습니다.

좋습니다.

vwx12345 committed:
refactor(permission): 활성 환경 변경

codingbaraGo reviewed (approved):
고생하셨습니다.
